go/private/go_repository.bzl

```diff
@@ -25,67 +25,84 @@ def _go_repository_impl(ctx):
   else:
     fail("neither commit or tag is specified", "commit")
 
-  # TODO(yugui): support submodule?
-  # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
-  remote = ctx.attr.remote
-  vcs = ctx.attr.vcs
-  importpath = ctx.attr.importpath
-  result = ctx.execute([
-      fetch_repo,
-      '--dest', ctx.path(''),
-      '--remote', remote,
-      '--rev', rev,
-      '--vcs', vcs,
-      '--importpath', importpath])
-  if result.return_code:
-    fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
+  if ctx.attr.url:
+    # explicit source url
+    if ctx.attr.vcs:
+        fail("cannot specify both of vcs and url")
+    ctx.download_and_extract(
+        url = ctx.attr.url,
+        sha256 = ctx.attr.sha256,
+        stripPrefix = ctx.attr.strip_prefix,
+        type = ctx.attr.type,
+    )
+  else:
+    # Using fetch repo
+    if ctx.attr.vcs and not ctx.attr.remote:
+      fail("if vcs is specified, remote must also be")
+    # TODO(yugui): support submodule?
+    # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
+    result = ctx.execute([
+        ctx.path(ctx.attr._fetch_repo),
+        '--dest', ctx.path(''),
+        '--remote', ctx.attr.remote,
+        '--rev', rev,
+        '--vcs', ctx.attr.vcs,
+        '--importpath', ctx.attr.importpath,
+    ])
+    if result.return_code:
+      fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
 
+  if (ctx.attr.disable_build_file_generation or 
+      ctx.path('BUILD').exists or 
```

Comment:
A couple of pathological cases might be worth considering:

  1. What if there's file named BUILD but it comes from the veeeerrry similar build system used by that other company? E.g. https://buckbuild.com/rule/go_binary.html or http://www.pantsbuild.org/go_readme.html?  This could be solved using a flag to _force_ the generation of a Bazel BUILD file, or BUILD.bazel as is typical of new_git_repository.

  2. What if Bazel BUILD files already exist but only in subdirectories of the base repo?  This may happen if my repo contains only packages that should be imported as, say, "github.com/jwpowell055/foo_repo/bar_package".